File: src/main/java/fr/axl/lvy/quote/QuoteService.java
package fr.axl.lvy.quote;

import fr.axl.lvy.documentline.DocumentLine;
import fr.axl.lvy.documentline.DocumentLineRepository;
import fr.axl.lvy.order.OrderA;
import fr.axl.lvy.order.OrderARepository;
import java.time.LocalDate;
import java.util.List;
import java.util.Optional;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class QuoteService {

  private final QuoteRepository quoteRepository;
  private final OrderARepository orderARepository;
  private final DocumentLineRepository documentLineRepository;

  QuoteService(
      QuoteRepository quoteRepository,
      OrderARepository orderARepository,
      DocumentLineRepository documentLineRepository) {
    this.quoteRepository = quoteRepository;
    this.orderARepository = orderARepository;
    this.documentLineRepository = documentLineRepository;
  }

  @Transactional(readOnly = true)
  public List<Quote> findAll() {
    return quoteRepository.findByDeletedAtIsNull();
  }

  @Transactional(readOnly = true)
  public Optional<Quote> findById(Long id) {
    return quoteRepository.findById(id);
  }

  @Transactional
  public Quote save(Quote quote) {
    return quoteRepository.save(quote);
  }

  @Transactional
  public void delete(Long id) {
    quoteRepository.findById(id).ifPresent(Quote::softDelete);
  }

  @Transactional
  public OrderA convertToOrderA(Quote quote, String orderNumber) {
    if (quote.getStatus() != Quote.QuoteStatus.ACCEPTED
        && quote.getStatus() != Quote.QuoteStatus.SENT) {
      throw new IllegalStateException("Quote cannot be converted in its current status");
    }

    var order = new OrderA(orderNumber, quote.getClient(), LocalDate.now());
    order.setQuote(quote);
    order.setClientReference(quote.getClientReference());
    order.setSubject(quote.getSubject());
    order.setBillingAddress(quote.getBillingAddress());
    order.setShippingAddress(quote.getShippingAddress());
    order.setCurrency(quote.getCurrency());
    order.setExchangeRate(quote.getExchangeRate());
    order.setIncoterms(quote.getIncoterms());
    order.setNotes(quote.getNotes());
    order.setConditions(quote.getConditions());
    order = orderARepository.save(order);

    var lines =
        documentLineRepository.findByDocumentTypeAndDocumentIdOrderByPosition(
            DocumentLine.DocumentType.QUOTE, quote.getId());
    for (var line : lines) {
      var newLine =
          new DocumentLine(DocumentLine.DocumentType.ORDER_A, order.getId(), line.getDesignation());
      newLine.setProduct(line.getProduct());
      newLine.setDescription(line.getDescription());
      newLine.setHsCode(line.getHsCode());
      newLine.setMadeIn(line.getMadeIn());
      newLine.setClientProductCode(line.getClientProductCode());
      newLine.setQuantity(line.getQuantity());
      newLine.setUnit(line.getUnit());
      newLine.setUnitPriceExclTax(line.getUnitPriceExclTax());
      newLine.setDiscountPercent(line.getDiscountPercent());
      newLine.setVatRate(line.getVatRate());
      newLine.setPosition(line.getPosition());
      newLine.recalculate();
      documentLineRepository.save(newLine);
    }

    quote.setStatus(Quote.QuoteStatus.ACCEPTED);
    quoteRepository.save(quote);

    order.recalculateTotals();
    return orderARepository.save(order);
  }
}

File: src/main/kotlin/fr/axl/lvy/quote/QuoteService.kt
package fr.axl.lvy.quote

import fr.axl.lvy.documentline.DocumentLine
import fr.axl.lvy.documentline.DocumentLineRepository
import fr.axl.lvy.order.OrderA
import fr.axl.lvy.order.OrderARepository
import java.time.LocalDate
import java.util.Optional
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class QuoteService(
  private val quoteRepository: QuoteRepository,
  private val orderARepository: OrderARepository,
  private val documentLineRepository: DocumentLineRepository,
) {
  @Transactional(readOnly = true)
  fun findAll(): List<Quote> = quoteRepository.findByDeletedAtIsNull()

  @Transactional(readOnly = true)
  fun findById(id: Long): Optional<Quote> = quoteRepository.findById(id)

  @Transactional fun save(quote: Quote): Quote = quoteRepository.save(quote)

  @Transactional
  fun delete(id: Long) {
    quoteRepository.findById(id).ifPresent { it.softDelete() }
  }

  @Transactional
  fun convertToOrderA(quote: Quote, orderNumber: String): OrderA {
    if (quote.status != Quote.QuoteStatus.ACCEPTED && quote.status != Quote.QuoteStatus.SENT) {
      throw IllegalStateException("Quote cannot be converted in its current status")
    }

    var order = OrderA(orderNumber, quote.client, LocalDate.now())
    order.quote = quote
    order.clientReference = quote.clientReference
    order.subject = quote.subject
    order.billingAddress = quote.billingAddress
    order.shippingAddress = quote.shippingAddress
    order.currency = quote.currency
    order.exchangeRate = quote.exchangeRate
    order.incoterms = quote.incoterms
    order.notes = quote.notes
    order.conditions = quote.conditions
    order = orderARepository.save(order)

    val lines =
      documentLineRepository.findByDocumentTypeAndDocumentIdOrderByPosition(
        DocumentLine.DocumentType.QUOTE,
        quote.id!!,
      )
    for (line in lines) {
      val newLine = DocumentLine(DocumentLine.DocumentType.ORDER_A, order.id!!, line.designation)
      newLine.product = line.product
      newLine.description = line.description
      newLine.hsCode = line.hsCode
      newLine.madeIn = line.madeIn
      newLine.clientProductCode = line.clientProductCode
      newLine.quantity = line.quantity
      newLine.unit = line.unit
      newLine.unitPriceExclTax = line.unitPriceExclTax
      newLine.discountPercent = line.discountPercent
      newLine.vatRate = line.vatRate
      newLine.position = line.position
      newLine.recalculate()
      documentLineRepository.save(newLine)
    }

    quote.status = Quote.QuoteStatus.ACCEPTED
    quoteRepository.save(quote)

    val orderLines =
      documentLineRepository.findByDocumentTypeAndDocumentIdOrderByPosition(
        DocumentLine.DocumentType.ORDER_A,
        order.id!!,
      )
    order.recalculateTotals(orderLines)
    return orderARepository.save(order)
  }
}
